Add emutls support to GC

jpf91 opened the pull request:
This adds support to make the GC scan TLS memory when emulated TLS is being used.

Generates one `__modtlsscan(scope void delegate(void* start, void* end) scan)` function per module. The function can be accessed from `ModuleInfo` and will be called from the garbage collector. The function must call `scan` for every TLS variable used in the module.

Example output (two distinct variables, unfortunately with the same name):

```
;; Function __modtlsscan (_D4test12__modtlsscanFMDFPvPvZvZv)
;; enabled by -tree-original

{
dg.func (dg.object, (void *) &tlsVar, (void *) (&tlsVar + 4));
dg.func (dg.object, (void *) &tlsVar, (void *) (&tlsVar + 4));
}
```

Also integrates nicely with gdb:

```
(gdb) break std.random.__modtlsscan
Breakpoint 1 at 0x114a72f: file ../../../../gcc-5-20150201/libphobos/src/std/random.d, line 1.
(gdb) run
Starting program: /mnt/hdd/GDC/objdir/x86_64-unknown-linux-gnu/libphobos/src/unittest 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".

Breakpoint 1, std.random.__modtlsscan(scope void(void*, void*) delegate) (dg=...)
    at ../../../../gcc-5-20150201/libphobos/src/std/random.d:1
```

There are some pros and cons with this approach:
- Should work everywhere (even native TLS targets should work)
- Requires no special libgcc support
- Should work just fine with all combinations of shared libraries
- Standard GCC EmuTLS so TLS variables can be used across language boundaries
- Probably slow with many TLS variables as the GC has to scan every variable individually
- Eager allocation & pressure on the emutls implementation: As all variables will be scanned in all threads, this will force all TLS variables to be allocated in all threads (Note: this is also the case for native TLS)

This passes the test suite so I'm confident it should work as expected. This still requires a review. For example I'm not sure if there's no better way to define `GNU_EMUTLS`. And then some code can probably be simplified by using some functions I didin't know ;-)

jpf91 committed:
Add GNU_EMUTLS version

jpf91 committed:
Emit __modtlsscan function

Generate one
__modtlsscan(scope void delegate(void* start, void* end) nothrow scan) nothrow
function per module. The function can be accessed from ModuleInfo and will be
called from the garbage collector. The function must call scan for every
TLS variable used in the module.

Example output:
--------------------
;; Function __modtlsscan (_D4test12__modtlsscanFMDFPvPvZvZv)
;; enabled by -tree-original

{
dg.func (dg.object, (void *) &tlsVar, (void *) (&tlsVar + 4));
dg.func (dg.object, (void *) &tlsVar, (void *) (&tlsVar + 4));
}
--------------------

There are some pros and cons with this approach:
+ Should work everywhere (even native TLS targets should work)
+ Requires no special libgcc support
+ Should work just fine with all combinations of shared libraries
+ Standard GCC EmuTLS, so TLS variables can be used across language
boundaries

- Probably slow with many TLS variables as the GC has to scan every
  variable individually
- Extern TLS variables are not scanned (could probably be fixed)
- Eager allocation: As all variables will be scanned in all
  threads, this will force all TLS variables to be allocated in
  all threads (Note: this is also the case for the native TLS
  implementation)

Regarding shared libraries:
This only requires that iterating over ModuleInfo really processes
every (active) module in every thread. The shared library support
already takes care of this ([1], [2], [3]). There's the limitation that
dlopened libraries are only visible in the loading thread & threads
started from that loading thread. But this is also true for TLS
ctors/dtors and ModuleInfo in general: Other threads don't run
TLS ctors/dtors for the loaded library and therefore cannot safely
access TLS anyway. So even if we can't scan the TLS data of the loaded
library in these threads, it shouldn't really matter as there should be
no live data in these cases.

jpf91 committed:
Runtime support for GNU_EMUTLS

jpf91 commented:
ping @ibuclaw

ibuclaw commented:
First pass at reading looks good to me.  How is the shared library support looking with this?

ibuclaw commented on libphobos/libdruntime/core/thread.d:
```diff
@@ -2640,7 +2647,19 @@ private void scanAllTypeImpl( scope ScanAllThreadsTypeFn scan, void* curStackTop
+                        m.scanTLS((p1, p2) => scan(ScanType.tls, p1, p2));
+                }
+            }
+            catch(Exception) {} // Can't throw as long as our delegate doesn't throw
```
Can't we just mark the delegate parameter as `nothrow` object.d?  Or is some fancy casting required here.

jpf91 replied:
We can and we should. But ModuleInfos `opApply` is not nothrow (Although it never throws as long as the foreach body doesn't throw) so the try/catch is still necessary.

jpf91 commented:
> How is the shared library support looking with this?

It should work fine. This only requires that iterating over `ModuleInfo` really processes every (active) module in every thread. The shared library support already takes care of this ([1], [2], [3]). There's the limitation that `dlopened` libraries are only visible in the loading thread & threads started from that loading thread. But this is also true for TLS ctors/dtors and `ModuleInfo` in general: Other threads don't run TLS ctors/dtors for the loaded library and therefore cannot safely access TLS anyway. So even if we can't scan the TLS data of the loaded library in these threads, it shouldn't really matter as there should be no live data in these cases. 

[1] https://github.com/D-Programming-Language/druntime/blob/master/src/rt/minfo.d#L275
[2] https://github.com/D-Programming-Language/druntime/blob/master/src/rt/sections_elf_shared.d#L51
[3] https://github.com/D-Programming-Language/druntime/blob/master/src/rt/sections_elf_shared.d#L271

jpf91 commented:
Updated:
- made delegate and `__modtlsscan` nothrow
- fixed whitespace issues & code formatting
- collect tls variables in `VarDeclaration::toObjfile` instead of `VarDeclaration::toSymbol`
- update changelog

ibuclaw commented:
I'd also be interested in some performance metrics if we did the same type of scan for native tls too.  As I understand it, not all platforms opt to use a packed section to store all tls data (Windows and OSX may be examples of this culprit).

ibuclaw commented:
I ask because we should really throw out `_tlsstart` and `_tlsend` too.

ibuclaw commented:
In any case, this looks good to me.  Merge when you're ready.

jpf91 commented:
> I'd also be interested in some performance metrics if we did the same type of scan for native tls too.

That's an interesting question. I'll write a small benchmark.

> As I understand it, not all platforms opt to use a packed section to store all tls data (Windows and OSX may be examples of this culprit).

Really? Doesn't performace suffer a lot in that case? I'd expect the TLS data per library & thread has to be in one block for a high performance implementation?

ibuclaw commented:
OK, had a small read up.  Windows apparently has a .tls section, but because of almost everything being a DLL, emulated tls via TlsAlloc/TlsGetValue seems to be the norm?
https://msdn.microsoft.com/en-gb/library/ms809762.aspx

jpf91 commented:
MS fixed that in windows vista:
https://msdn.microsoft.com/en-us/library/2s9wt68x.aspx
(last point)
http://www.nynaeve.net/?p=189

jpf91 commented:
Some test results:
bench.d: http://dpaste.dzfl.pl/f2428cc99d8c
bench2.d: http://dpaste.dzfl.pl/7af8477e0b6e
Place in same folder, compile & run bench2.d

With 50 modules, every module with 1000 TLS vars:
1 thread: 8x slower
500 threads: 10x slower
5000 threads: 16x slower (~4s instead of 0,25s)

With 50 modules, every module with 100 TLS vars:
1 thread: 2x slower
500 threads: 3x slower
5000 threads: 5x slower

With 50 modules, every module with 10 TLS vars:
~1,5x slower

So it's a significant performance hit.

jpf91 commented:
@ibuclaw BTW: I think I found a bug in the backend. I somehow missed this a first, but `runable/mangle.d` doesn't compile with emutls. `set_user_assembler_name` internally prepends a `*` to the name. This is then later removed when emitting the variable. But when the emutls names are constructed, it doesn't strip the `*`. The name ends up like `emutls_v*name` and fails to assemble. This only affects TLS variables with `pragma(mangle)`. Do you think I should report this on the GCC bugtracker?
